main/src/ca/uwaterloo/flix/language/phase/Enumgen.scala

```diff
@@ -0,0 +1,353 @@
+package ca.uwaterloo.flix.language.phase
+
+import org.objectweb.asm
+import org.objectweb.asm.{ClassWriter, Label}
+import org.objectweb.asm.Opcodes._
+import ca.uwaterloo.flix.language.ast.{Type, _}
+import ca.uwaterloo.flix.util.InternalCompilerException
+
+/**
+  * Created by ramin on 2017-05-22.
+  *
+  * Generates classes and interfaces which will represent enums
+  */
+object Enumgen {
+  import CodegenHelper._
+
+  /**
+    * Generates an interface for TagInterface which includes one method which
+    * returns the name of the enum case which implements this interface
+    * @param name name of the class
+    * @return byte code representation of the class
+    */
+  def compileTagInterface(name: List[String]) : Array[Byte] = {
+    val visitor = new ClassWriter(0)
+
+    visitor.visit(V1_8, ACC_PUBLIC + ACC_ABSTRACT + ACC_INTERFACE, decorate(name), null,
+      asm.Type.getInternalName(Constants.objectClass), null)
+
+    val getTagMethod = visitor.visitMethod(ACC_PUBLIC + ACC_ABSTRACT, "getTag",
+      s"()${asm.Type.getDescriptor(Constants.stringClass)}", null, null)
+    getTagMethod.visitEnd()
+
+    visitor.visitEnd()
+    visitor.toByteArray
+  }
+
+  /**
+    * Generates an interface for each enum.
+    * Each case of the enum implements this interface. This interface extends `TagInterface`.
+    * @param name name of the class
+    * @param tagInterface the tagInterface
+    * @return byte code representation of the class
+    */
+  def compileEnumInterface(name: List[String], tagInterface: Class[_]): Array[Byte] = {
+    val visitor = new ClassWriter(0)
+
+    visitor.visit(V1_8, ACC_PUBLIC + ACC_ABSTRACT + ACC_INTERFACE, decorate(name), null,
+      asm.Type.getInternalName(Constants.objectClass), Array(asm.Type.getInternalName(tagInterface)))
+
+    visitor.visitEnd()
+    visitor.toByteArray
+  }
+
+  /**
```

Comment:
It might be useful to give en example as part of the comment.

E.g. given an enum of this structure (you can use list) then this method generates the following class (shown in Java code)

Same goes for every other method.